docs/sql/02-type-system.md

```diff
@@ -0,0 +1,175 @@
+# SQL Type System
+
+## Overview
+
+The type system defines how objects of different types interact with each other in the Hazelcast Mustang engine.
+The type system is defined by the list of supported types, type mapping, and type conversion rules.
+
+A type is defined by name, precedence, precision, and scale.
+
+- **Name** is a textual representation of type name
+- **Precision** is the total count of significant digits in the whole number, applicable to **numeric types**
+- **Precedence** is a comparable value that is used for type inference in expressions. A type with a
+  higher value has precedence over a type with a lower value. If two types have the same precedence value, then
+  the type with higher precision has precedence.
+
+**Type family** is a collection of types with the same name, but different precisions. All types
+within a family have the same name and precedence. For example, `INTEGER(11)` and `INTEGER(12)` are two types
+from the same `INTEGER` family.
+
+The scale is not used in Hazelcast Mustang. It is applicable only for `DECIMAL`, `REAL`, and `DOUBLE` types.
+Instead of defining it as a separate value, we just treat these types as types with infinite scale.
+
+## 1 Supported Types
```

Comment:
entire chapter misses `JSON` mapped to `HazelcastJsonValue`

Reply:
Good catch, I guess, it's a point for a separate PR.

Reply:
maybe better to do it in this PR, because you already added JSON to the big table, but without conversions to VARCHAR